presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -0,0 +1,283 @@
+                        aggregationNode,
+                        joinSource.getCriteria(),
+                        ImmutableList.<Symbol>builder().addAll(rewrittenLeft.getOutputSymbols()).addAll(aggregationNode.getOutputSymbols()).build(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+
+                return coalesceWithNullAggregation(aggregationNode, joinNode);
+            }
+
+            if (joinSource.getType() == JoinNode.Type.RIGHT
+                    && !joinSource.getFilter().isPresent()
+                    && groupsOnAllOuterTableColumns(node, joinSource)
+                    && isDistinct(joinSource.getRight())) {
+                List<Symbol> groupingKeys = joinSource.getCriteria().stream()
+                        .map(JoinNode.EquiJoinClause::getLeft)
+                        .collect(toImmutableList());
+                AggregationNode aggregationNode = new AggregationNode(
+                        node.getId(),
+                        context.defaultRewrite(joinSource.getLeft()),
+                        node.getAggregations(),
+                        node.getFunctions(),
+                        node.getMasks(),
+                        ImmutableList.of(groupingKeys),
+                        node.getStep(),
+                        node.getHashSymbol(),
+                        node.getGroupIdSymbol());
+                PlanNode rewrittenRight = context.defaultRewrite(joinSource.getRight());
+                JoinNode joinNode = new JoinNode(
+                        joinSource.getId(),
+                        joinSource.getType(),
+                        aggregationNode,
+                        rewrittenRight,
+                        joinSource.getCriteria(),
+                        ImmutableList.<Symbol>builder().addAll(aggregationNode.getOutputSymbols()).addAll(rewrittenRight.getOutputSymbols()).build(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+                return coalesceWithNullAggregation(aggregationNode, joinNode);
+            }
+            return context.defaultRewrite(node, context.get());
+        }
+
+        // When the aggregation is done after the join, there will be a null value that gets aggregated over
+        // where rows did not exist in the inner table.  For some aggregate functions, such as count, the result
+        // of an aggregation over a single null row is one or zero rather than null. In order to ensure correct results,
+        // we add a coalesce function with the output of the new outer join and the agggregation performed over a single
+        // null row.
+        private PlanNode coalesceWithNullAggregation(AggregationNode aggregationNode, PlanNode outerJoin)
+        {
+            NullLiteral nullLiteral = new NullLiteral();
```

Comment:
I find this whole section somewhat hard to read. I'd suggest breaking it up into smaller sections that are distinct from each other. I'll play with the code a bit to see if I can make a concrete suggestion.